### ISSUE TITLE
Question about ZenPhoto20

### ISSUE THREAD

**Issue by mebels** [ZenPhoto20/ZenPhoto20#731]
I have a few questions about ZenPhoto20.

Currently I use Gallery3 for some years. At that time, I switched from a pure HTML gallery to Gallery3 because I liked it the most. Back then, I also tested JoomGallery, Piwigo, Coppermine and Zenphoto.
What I did not like about Zenphoto was this "on the fly" image processing and the cache. It's more neatly designed in Gallery3. What I really need is the tag function. In addition, the URLs in Gallery3 can be tree-structured, without file extension. Unfortunately, Gallery3 went into hibernation a short time later.
I only made a few small core hacks. Mainly I've got the responsive theme working, but only for me, with many adjustments. And I think my photo gallery website is good to find and well visited.

For some days I test ZenPhoto again and come back to the difficulties. Mainly URL, "on the fly" and /cache.

I like it more like Gallery3.
/albums
/resize
/thumbs
(possibly for ZenPhoto: backend-thumbs)
And move all images with moving albums/images (in /albums, /resize, /thumbs, ...)

My questions to ZenPhoto20:

1. Does ZenPhoto20 use the "on the fly" function to create images?
2. Does ZenPhoto20 use the folder /cache to save the resize images and does ZenPhoto20 save all resize and thumbs in the same subfolder?
I like it more 
3. Are these image pages URLs possible in ZenPhoto20?
```https://www.example.com/album/subalbum/IMG0001```
4. Is it possible to incorporate a responsive basic theme here? (I created one).

**Comment by sbillard** [comment]
1. As I recall, Gallery stored the images in the database. That was the original reason I did not go with it. Zenphoto is file system based. This means that there may appear new images without ZenPhoto20 having any prior warning. With Gallery if you did not use its tool to upload the image then the image simply did not exist. Aside from the waste of database resources, this scheme was (to me) too cumbersome in operation.

ZenPhoto20 provides a utility to create these images by request so you can anticipate the need and have the image ready. You can invoke this function after you upload your images and all will be present ahead of time. Most people do not bother with this, though. Normally the next thing you do after uploading images is to view them in your theme. At that point the cached image is created. Most think this is a much more convenient workflow. 

I really wonder why you ask this question. Certainly if an image of the right size does not exist when a web page is accessed the software must either create it or the page will show a "broken" image. I would think that the former is preferable. 

2. ZenPhoto20 does not organize the various resized images into subfolders. This has never before been cited as an issue. Can you please explain your reasoning as to why this is a problem? These images are not normally viewed directly, they are loaded onto web pages where the image URL reference would normally never be seen by the visitor. If there is a compelling reason for subfolder organization it will be a simple addition. However, note that it is not always possible to identify what resized images are used as thumbnails and which are simply small sized images. Unless the theme actually requests the image as a thumbnail it will be considered a normal sized image.

3. Currently the _unique images_ option is disabled if there is no mod_rewrite_suffix. I think that will not be a problem to change and could be a good enhancement. On the other hand, having the link use a "standard" web page suffix is better for SEO engines. Links without a suffix are normally considered to be to folders. I would really recommend you use a mod_rewrite_suffix such as ".htm" or even ".php".

4. I presume that your theme will work with ZenPhoto20 if it works worth zenphoto. If not we can correct either the theme or the ZenPhoto20 software. There is both a compatibility plugin and a migration assistance tool for themes to make them compatible to ZenPhoto20. The compatibility plugin should not be used in the long term for performance reasons. (You will find ZenPhoto20 a better performer than its predecessor) so the migration is recommended. The migration tool has been successfully used on many themes. For instance, all the "extra" themes in ZenPhoto20 were third party created themes that were migrated to ZenPhoto20 from their original zenphoto sources.

**Comment by sbillard** [comment]
I have done some investigation into leaving out the "suffix" for the _unique image_ option. I think it is something that we will NOT do. The primary reason for this is that it would seriously impact performance. As noted above, URL links without a suffix are normally considered to a base folder and by default the _index.htm_, etc.) file therein is loaded. Within the context of ZenPhoto20 this would imply a link to an album, not an image. While it would be possible to discover if there were an image of the same base name, it does add overhead to EVERY album link. So considering it is not a good SEO option as well I think it would be a poor choice to implement. Instead you should choose a rewrite suffix of your preference.

**Comment by mebels** [comment]
Because of the "on the fly": ZenPhoto tells me that it does not move the images in the folder cache (and its subfolders) when moving the images in the backend. OK, that would be synonymous with "all resizes and Thumbs in a folder" possible. I find it only clearer to use different folders. But with the "on the fly" function it is actually a bit different, because you do not have to check if all resizes and thumbs are present. Unfortunately, I have no argument that can convince you. Except that renaming the resizes is then not necessary and making comparison between original, resize and thumbs easier (if without "on the fly").
I have only one argument: with "on the fly" all images are saved after a short time, why not immediately?

1. different resize folder: no rename.
2. resize / thumbs save immediately: "on the fly" stores anyway after a short time all.

A compromise:
You have described that "many" do not care about saving resize and thumbs immediately. OK. But that would be important to me. Since I edit each image after the "put online" (adding title, description and tags), it would be enough if the resizes and thumbs were saved immediately when calling the "edit images" backend webpage. With this call of the "edit images" backend webpage, the thumbs for the backend will be created anyway. Maybe this could be inserted as an option, not as a standard? I would be happy about that.


Because of the URL: You only almost correctly recognized this with "index.php". If a URL does not have a suffix AND a closing slash "/" then it is assumed to be a folder and the "index.php" is loaded. But this is only true for file system based URLs (ie in pure HTML), not PHP generated content. (Remembering PHP URLs with query string "eample.com/index.php? ..." and hiding "index.php" using htaccess.)
The URLs to each subpage in a folder in a file based structur have different filenames. This is also not true here, since the page structur is not file based but generates PHP.

In the best case each page has its own folder and a (hiding) index.php. This can be assumed if every page is loaded without a suffix. But, missing the closing slash "/". So URLs without suffix ".php" and without closing slash "/" are recognized as "files" with hiding "suffix". Hiding "suffix" is nothing bad and may even be a precautionary measure in CSM systems, or not necessary in CMS systems. I am not aware that search engines attach particular importance to the fact that the address must be marked as a "real" webpage as a "file". In CMS systems, this is nonsense. In CMS systems, a URL is just an address, not a pointer. However, it is probably problematic if a wrong suffix is used (as in Zenphoto the images file extension). This can be recognized as a pointer, which does not point to an image file, but a (PHP generated) HTML file. So that could be classified as a fake. ZenPhoto admits that too.

You have to distinguish two different things here.

File based: image files.
Not file based: pages.

It is thus "obsolete", or only necessary for HTML or file system based web pages in tree structure.

The problem with files in a folder with the same name, like
IMG0001.JPG
img0001.jpg
IMG0001.PNG
...

Automatic Renaming (as in Gallery3) I do not like, like "img0001-2.jpg".
My suggested solution would be to include a detection feature that detects if "duplicate" file names exist. If "duplicate" file names exist, then all files with "duplicate" file names will be displayed on a summary page, as in a search result page when a URL to "duplicate" filename is call.

As far as my basic responsive theme is concerned. How can I best show this? I do not have that much experience from github yet. I would like to contribute.

**Comment by sbillard** [comment]
I guess that a file based system like ZenPhoto20 and its predecessor is then just not for you. I suggest you continue your search for the photo gallery that meets your expectations as clearly the architecture that we have designed does not.

**Comment by mebels** [comment]
I've tested all the available ones several times and intensively over several years. Coppermine, Piwigo, JoomGallery, Gallery3, (Koken), ZenPhoto, ZenPhoto20. I chose Gallery3. But even at Gallery3, not everything was as standard as I wanted. A few small core hacks were necessary. And with ZenPhoto just a few core hacks are necessary. At least with the first ZenPhoto Core hack, which I have created successfully today, it looks very good. Thanks for the suggestions.
